lms/services/canvas.py

```diff
@@ -1,85 +1,119 @@
 
-        return self.api.public_url(file_id)
+        :raise CanvasFileNotFoundInCourse: if check_in_course was True and the
+            current user can't see file_id in course_id's list of files
 
-    def can_see_file_in_course(self, file_id: str, course_id: str) -> bool:
-        """
-        Return True if the current user can see file_id in course_id.
-
-        Return False if the current user cannot currently see a file with ID
-        file_id in the course with ID course_id. This could be because the file
-        is in another course, because the file was in course_id but has been
-        deleted, or because the file is in course_id but the current user
-        doesn't have permission to see it (for example files that are marked as
-        "unpublished" in Canvas can only be seen by teachers, not students).
+        :raise CanvasAPIPermissionError: if the user gets a permissions error
+            from the Canvas API when trying to get a public URL for file_id
         """
-        files_in_course = self.api.list_files(course_id)
-
-        for file_ in files_in_course:
-
+        # If there's a previously stored mapping for file_id use that instead.
+        effective_file_id = module_item_configuration.get_canvas_mapped_file_id(file_id)
+
+        try:
+            if check_in_course:
+                self._finder.assert_file_in_course(course_id, effective_file_id)
+            return self.api.public_url(effective_file_id)
+        except (CanvasFileNotFoundInCourse, CanvasAPIPermissionError):
+            # The user can't see the file in the course. This could be because:
+            #
+            # * The course has been copied so the assignment's file_id is from
+            #   the original course (and the user isn't in the original course)
+            # * The assignment's file has been deleted from Canvas
+            # * The file *is* in the course but the user can't see it
+            #   (only instructors can see "unpublished" files in Canvas)
+            #
+            # We'll try to find another copy of the same file that the current
+            # user *can* see in the current course and use that instead.
+            found_file_id = self._finder.find_matching_file_in_course(
+                course_id, effective_file_id
+            )
+
+            if not found_file_id:
+                raise
+
+            # Try again to return a public URL, this time using found_file_id.
+            url = self.api.public_url(found_file_id)
+
+            # Store a mapping so we don't have to re-search next time.
+            module_item_configuration.set_canvas_mapped_file_id(file_id, found_file_id)
+
+            return url
+
+
+class CanvasFileFinder:
+    """A helper for finding file IDs in the Canvas API."""
+
+    def __init__(self, canvas_api, file_service):
+        self._api = canvas_api
+        self._file_service = file_service
+
+    def assert_file_in_course(self, course_id: str, file_id: str) -> bool:
```

Comment:
```suggestion
    def assert_file_in_course(self, course_id: str, file_id: str):
```

I don't think this returns a bool anymore? It just raises or returns?

Reply:
I just got rid of all the annotations